packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php

```diff
@@ -147,44 +144,58 @@ private function processNodesWithMixinHandling(
     /**
      * @param Node[] $nodes
      */
-    private function isMixinInSource(array $nodes, SmartFileInfo $smartFileInfo): bool
+    private function isMixinInSource(array $nodes): bool
     {
-        $realPath = $smartFileInfo->getRealPath();
-        return (bool) $this->betterNodeFinder->findFirst($nodes, function (Node $node) use ($realPath): bool {
-            if (! $node instanceof FullyQualified) {
+        return (bool) $this->betterNodeFinder->findFirst($nodes, function (Node $node): bool {
+            if (! $node instanceof FullyQualified && ! $node instanceof Class_) {
                 return false;
             }
 
-            $className = $node->toString();
-
-            // fix error in parallel test
-            // use function_exists on purpose as using reflectionProvider broke the test in parallel
-            if (function_exists($className)) {
+            if ($this->classAnalyzer->isAnonymousClass($node)) {
                 return false;
             }
 
-            $hasClass = $this->reflectionProvider->hasClass($className);
+            $className = $node instanceof FullyQualified ? $node->toString() : $node->namespacedName->toString();
 
-            if (! $hasClass) {
-                return false;
-            }
+            return $this->isCircularMixin($className);
+        });
+    }
 
-            $classReflection = $this->reflectionProvider->getClass($className);
-            if ($classReflection->isBuiltIn()) {
+    private function isCircularMixin(string $className): bool
+    {
+        // fix error in parallel test
+        // use function_exists on purpose as using reflectionProvider broke the test in parallel
+        if (function_exists($className)) {
+            return false;
+        }
+
+        $hasClass = $this->reflectionProvider->hasClass($className);
+
+        if (! $hasClass) {
+            return false;
+        }
+
+        $classReflection = $this->reflectionProvider->getClass($className);
+        if ($classReflection->isBuiltIn()) {
+            return false;
+        }
+
+        foreach ($classReflection->getMixinTags() as $mixinTag) {
+            $type = $mixinTag->getType();
+            if (!$type instanceof ObjectType){
                 return false;
             }
 
-            $fileName = $classReflection->getFileName();
-            if ($fileName === false) {
-                return false;
+            if ($type->getClassName() === $classReflection->getName()) {
```

Comment:
please assign classReflection->getName() to variable before loop to improve performance